e2e-federation-tests/infrastructure/orders/api/types.py

```diff
@@ -0,0 +1,35 @@
+from typing import List, Optional
+
+import strawberry
+from strawberry import ID
+
+
+@strawberry.type
+class Order:
+    id: ID
+    customer_name: str
+
+
+@strawberry.federation.type(keys=["id"], extend=True)
+class Product:
+    id: ID = strawberry.federation.field(external=True)
+    name: str = strawberry.federation.field(external=True)
+
+    @strawberry.federation.field(requires=["name"])
+    def code(self) -> str:
+        return f"OrderService:{self.name}"
+
+    @strawberry.field
+    def orders(self) -> List[Order]:
+        if self.id == "1":
+            return [Order(id=ID("1"), customer_name="Marco")]
+        elif self.id == "2":
+            return [Order(id=ID("2"), customer_name="Patrick")]
+        elif self.id == "500":
+            return [Order(id=ID("3"), customer_name="Ethan Winters")]
+        else:
+            return []
+
+    @classmethod
+    async def resolve_reference(cls, id: ID, name: Optional[str] = None):
+        return cls(id=id, name=name)  # type: ignore
```

Comment:
`type: ignore` because mypy fails because `name` is Optional in `resolve_reference` but it can't be in the Product fields list because it needs to be the same type as on GraphQL

The reason `name` is Optional and by default `= None` is that the Gateway will pass the `name` only if the client asked for the `code` field (i.e the field that `@requires` it)

Reply:
Can this comment be added w/in the code?

Reply:
sure, will do it (this code is now in the other repo)

Reply:
Have a look here: https://github.com/strawberry-graphql/e2e-federation/pull/1/files